### ISSUE TITLE
[apex] string variable with bind operator causes false positive for ApexSOQLInjection

### ISSUE THREAD

**Issue by SeanWhalenUKG** [pmd/pmd#3696]
**Affects PMD Version:** 
 2.12.0
**Rule:**
ApexSOQLInjection
 [SOQL Injection](https://pmd.github.io/latest/pmd_rules_apex_security.html#apexsoqlinjection)

**Description:**
The sample code below uses a colon (":") as a binding operator to add the value of the variable to the query. Salesforce documents binding as a defense against SOQL injection. See:

[…With a bind variable, the attacker isn’t able to break out and control the SOQL query.](https://trailhead.salesforce.com/en/content/learn/modules/secure-serverside-development/mitigate-soql-injection)

and

[…The variable prepended with a colon is known as a bind variable. This tells the database layer, in this case SOQL, to treat everything in that variable as data, no matter what characters are in it. This means that no matter what the user types in, they can’t break out of the intended behavior of the query and manipulate the query](https://developer.salesforce.com/docs/atlas.en-us.secure_coding_guide.meta/secure_coding_guide/secure_coding_sql_injection.htm)

**Code Sample demonstrating the issue:**

```
    public static List<contact> getContacts(List<String> recordIds) {
        String queryString = 'SELECT Id, Name '+
            ' FROM contact'+
            ' WHERE ID IN :recordIds';
       
        List<contact> records =  Database.query(queryString);
        return records;
    }  

```

**Expected outcome:**
    No violation

PMD reports a SOQLInjection risk at the Database.query line, but that's wrong. That's a false positive.

**Comment by m0rjc** [comment]
I wonder if this is due to the concatenations, not the bind.
The code sees that it's a String and it's not directly derived from a Literal.

I've started looking at variables in #4650 but that is turning into quite a large problem.

**Comment by rcorfieldffdc** [comment]
I wonder if this is due to the concatenations, not the bind.
The code sees that it's a String and it's not directly derived from a Literal.

I've started looking at variables in https://github.com/pmd/pmd/issues/4650 but that is turning into quite a large problem.

**Comment by jsotuyod** [comment]
Indees, the issue is not the binds, but the concatenation. This is a duplicate of #2955
